jgrapht-core/src/main/java/org/jgrapht/graph/GraphWalk.java

```diff
@@ -174,6 +201,229 @@ public String toString()
+     * @return a reversed GraphWalk
+     */
+    public GraphWalk<V,E> reverse(Function<GraphWalk<V,E>,Double> walkWeightCalculator) throws InvalidGraphPathException {
+        List<V> revVertexList = null;
+        List<E> revEdgeList = null;
+
+        if(vertexList != null) {
+            revVertexList = new ArrayList<>(this.vertexList);
+            Collections.reverse(revVertexList);
+
+            //Check validity of the path. If the path is invalid, then calculating its weight may result in an undefined exception.
+            //If an edgeList is provided, then this check can be postponed to the construction of the reversed edge list
+            if(!graph.getType().isUndirected() && edgeList == null) {
+                for (int i = 0; i < revVertexList.size() - 1; i++) {
+                    V u = revVertexList.get(i);
+                    V v = revVertexList.get(i + 1);
+                    if (graph.getEdge(u, v) == null)
+                        throw new InvalidGraphPathException("this walk cannot be reversed. The graph does not contain a reverse arc for arc " + graph.getEdge(v, u));
+                }
+            }
+        }
+
+        if(edgeList != null){
+            revEdgeList = new ArrayList<>(this.edgeList.size());
+
+            if(graph.getType().isUndirected()){
+                revEdgeList.addAll(this.edgeList);
+                Collections.reverse(revEdgeList);
+            }else{
+                ListIterator<E> listIterator = this.edgeList.listIterator(edgeList.size());
+                while (listIterator.hasPrevious()) {
+                    E e = listIterator.previous();
+                    V u = graph.getEdgeSource(e);
+                    V v = graph.getEdgeTarget(e);
+                    E revEdge=graph.getEdge(v, u);
+                    if(revEdge == null)
+                        throw new InvalidGraphPathException("this walk cannot be reversed. The graph does not contain a reverse arc for arc "+e);
+                    revEdgeList.add(revEdge);
+                }
+            }
+        }
+        //Update weight of reversed walk
+        GraphWalk<V,E> gw=new GraphWalk<>(this.graph, this.endVertex, this.startVertex, revVertexList, revEdgeList, 0);
+        gw.setWeight(walkWeightCalculator.apply(gw));
+        return gw;
+    }
+
+    /**
+     * Concatenates the specified GraphWalk to the end of this GraphWalk. This action can only be performed if the end
+     * vertex of this GraphWalk is the same as the start vertex of the extending GraphWalk
+     * @param extension GraphPath used for the concatenation.
+     * @param walkWeightCalculator Function used to calculate the weight of the GraphWalk obtained after the concatenation.
+     * @return a GraphWalk that represents the concatenation of this object's walk followed by the walk specified in the
+     * extension argument.
+     */
+    public GraphWalk<V,E> concat(GraphPath<V,E> extension, Function<GraphWalk<V,E>,Double> walkWeightCalculator){
```

Comment:
I'm not entirely sure I understand your 2nd comment. What exactly do you propose to add? In your BiFunction, what are the first 2 double arguments?

Simply combining the 2 weights, e.g. taking the sum of the 2 paths can currently be done like this:
`GraphWalk<Integer, DefaultEdge> gw3=gw1.concat(gw2, gw-> gw1.getWeight()+gw2.getWeight());`

I'm using the walkWeightCalculator here for the following reason. Given two paths g1, g2, where the weight of the path is determined by the value of its vertices. If g1=[0,1,2], g2=[2,3,4,5], g3=g1.concat(g2), then the weight w(g3) != w(g1)+w(g2) because this would count the value of vertex 2 twice. Therefore it's useful to have the walkWeightCalculator here.